TESTING.asciidoc

```diff
@@ -76,7 +76,177 @@ In order to start a node with a trial license execute the following command:
 This enables security and other paid features and adds a superuser with the username: `elastic-admin` and
 password: `elastic-password`.
 
-==== Other useful arguments
+
+== Running verification tasks
+
+To run all verification tasks, including static checks, unit tests, and integration tests:
+
+---------------------------------------------------------------------------
+./gradlew check
+---------------------------------------------------------------------------
+
+Note that this will also run the unit tests and precommit tasks first. If you want to just
+run the integration tests (because you are debugging them):
+
+---------------------------------------------------------------------------
+./gradlew integTest
+---------------------------------------------------------------------------
+
+If you want to just run the precommit checks:
+
+---------------------------------------------------------------------------
+./gradlew precommit
+---------------------------------------------------------------------------
+
+Some of these checks will require `docker-compose` installed for bringing up
+test fixtures. If it's not present those checks will be skipped automatically.
+
+== How to write good tests?
+
+=== Test pyramid
+
+-------------------------------------------------
+                  /\
+                 /  \
+                /    \
+               /manual\
+              /________\ 
+             /          \
+            / 3rd party  \ 
+           /      OS      \   
+          /      BWC       \   
+         / Internal Cluster \ 
+        /  External (REST)   \ 
+       /        Unit          \
+      /________________________\
+-------------------------------------------------
+
+As a rule of thumb, always prefer writing tests at the lowest possible level 
+of the pyramid. 
+
+==== Unit Tests 
+
+Unit tests are the preferred way to test some functionality: most of the time
+they are simpler to understand, more likely to reproduce, and unlikely to be
+affected by changes that are unrelated to the piece of functionality that is
+being tested.
+
+**`ESTestCase`**: The base class of all tests. 
```

Comment:
This isn't exactly correct. We have some test cases that extend LuceneTestCase directly. Maybe say "most" instead of "all"? I would also make the distinction that there are many intermediate base test classes extending from this.